dask/array.py

```diff
@@ -240,3 +241,299 @@ def concatenate(arrays, axis=0):
+        slice is not 0?
+      i.e. should we return
+        {(y,0):(get, (x,1), slice(3,5))}
+      or
+        {(y,1):(get, (x,1), slice(3,5))}
+
+    - Can we always assume that block (x,1) is the second block? yes, per conversation with @mrocklin.
+    - We also assume that the first slice is (out_name, 0), even if we
+        slice from an inner block like (in_name, 3)
+    - How do we handle a case where the slice size is greater than
+        the block size?
+                       
+    """
+
+    #Step 0: Setup all the variables we will need for the actual slicing
+    #  calculations. This includes the proper starting and stopping
+    #  indexes and step size. Remember that all of these can be negative,
+    #  and all of them can be None
+    start = slice_spec.start
+    stop = slice_spec.stop
+    step = slice_spec.step
+    
+    if start < 0:
+        start = shape[0]+start
+    if stop < 0:
+        stop = shape[0]+stop
+
+    if start is None:
+        start = 0
+    if stop is None:
+        stop = shape[0]
+    if step is None:
+        step = 1
+
+    if start >= shape[0] or start == stop:
+        return {}
+        
+    if stop > shape[0]:
+        stop = shape[0]
+                        
+
+    #Step 1: All of our input variables should now be setup correctly,
+    #  so figure out where to start and stop based on the blocksize
+    #  of the input data
+    leftmost_block = start/blockshape[0]
+    #We subtract 1 from stop so that if the rightmost index
+    #falls on a block boundary, we won't include a slice in the next block
+    #  i.e. slice(0,20) on blockshape(20,) should give us
+    #  (x,0), slice(0,20)
+    #  NOT
+    #  ((x,0), slice(0,20)), ((x,1), slice(0,0))
+    rightmost_block = (stop-1)/blockshape[0]
+    leftmost_block_start_index = start % blockshape[0]
+    rightmost_block_stop_index = stop - (rightmost_block*blockshape[0])
+    num_blocks = rightmost_block - leftmost_block
+
+    
+        
+    dask = {}
+    #There are three main cases to account for when doing the slicing
+    #  within a dask graph
+    #1. slicing within a single block
+    #2. slicing with only adjacent blocks, i.e. blocks 1,2
+    #3. slicing with multiple contiguous blocks, i.e. block 1,2,3,4
+    if leftmost_block == rightmost_block:    
+        dask[(out_name, 0)] = (operator.getitem, (in_name, leftmost_block), slice(leftmost_block_start_index, rightmost_block_stop_index))
+    else:
+        dask[(out_name, 0)] = (operator.getitem, (in_name, leftmost_block), slice(leftmost_block_start_index, blockshape[0]))
+        dask[(out_name, num_blocks)] = (operator.getitem, (in_name, rightmost_block), slice(0, rightmost_block_stop_index))
+            
+    for block_num in range(1, num_blocks):
+        dask[(out_name, block_num)] = (operator.getitem, (in_name, leftmost_block+block_num), slice(0, blockshape[0]))
+
+    return dask
+
+
+
+def _slice_1d(dim_shape, blocksize, index):
+    """
+    Parameters
+    ----
```

Comment:
Style nit, should extend `----` further to cover all of the text `Parameters`.  Same as below.